Add accessibility announcements to a11y_assessments's SnackBars

nancyflutter opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Part of https://github.com/flutter/flutter/issues/173065

### What was changed:
- **Added accessibility announcements** to SnackBar content. Now, when a SnackBar appears, its content will be announced for screen reader users using `SemanticsService`.

### Why this is important:
- **Accessibility**: Ensures users with visual impairments can hear SnackBar content, enhancing the user experience for a broader audience.

### Related Issue:
- Fixes issue [#173065](https://github.com/flutter/flutter/issues/173065) regarding accessibility for SnackBars.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

nancyflutter committed:
Fix: Announce Snackbar message for accessibility compliance

nancyflutter committed:
Fix: Announce Snackbar message for accessibility compliance

chunhtai commented on dev/a11y_assessments/lib/use_cases/snack_bar.dart:
```diff
@@ -29,6 +30,22 @@ class MainWidget extends StatefulWidget {
+    );
+
+    Future<void>.delayed(const Duration(milliseconds: 100), () {
+      debugPrint('Announcing: $message');
```
let's remove print statement

chunhtai commented on dev/a11y_assessments/lib/use_cases/snack_bar.dart:
```diff
@@ -29,6 +30,22 @@ class MainWidget extends StatefulWidget {
+      ),
+    );
+
+    Future<void>.delayed(const Duration(milliseconds: 100), () {
```
we should only do this for MediaQueryData.supportsAnnounce is true, you can get this from MediaQuery.of(context).

chunhtai commented on dev/a11y_assessments/test/snack_bar_test.dart:
```diff
@@ -7,21 +7,25 @@ import 'package:flutter_test/flutter_test.dart';
     await pumpsUseCase(tester, SnackBarUseCase());
-    final Finder findHeadingLevelOnes = find.bySemanticsLabel('SnackBar Demo');
     await tester.pumpAndSettle();
+    final Finder findHeadingLevelOnes = find.bySemanticsLabel('SnackBar Demo');
```
Can you add a expect that the message is announced? See this file for example
https://github.com/flutter/flutter/blob/master/packages/flutter/test/semantics/semantics_service_test.dart

nancyflutter committed:
Fix: Announce Snackbar message for accessibility and Implemented widget tests to verify Snackbar visibility and announcement.

nancyflutter committed:
Fix: Add accessibility announcement when showing a SnackBar.

chunhtai commented on dev/a11y_assessments/test/snack_bar_test.dart:
```diff
@@ -3,25 +3,40 @@
+    });
   });
-}
+}
```
```suggestion
}

```

chunhtai committed:
Update dev/a11y_assessments/test/snack_bar_test.dart

chunhtai reviewed (approved):
LGTM

chunhtai committed:
Merge branch 'master' into snackbar_a11y_fix

loic-sharma commented on dev/a11y_assessments/lib/use_cases/snack_bar.dart:
```diff
@@ -29,6 +30,19 @@ class MainWidget extends StatefulWidget {
+
+    final MediaQueryData mediaQuery = MediaQuery.of(context);
+    if (mediaQuery.supportsAnnounce) {
+      WidgetsBinding.instance.addPostFrameCallback((_) {
+        SemanticsService.announce(message, TextDirection.ltr);
```
See: https://github.com/flutter/flutter/pull/172669

nancyflutter committed:
Update dev/a11y_assessments/lib/use_cases/snack_bar.dart

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

matanlurey reviewed (changes requested):
Just a note - primarily to the _reviewers_: Aside from the CLA issue, this PR doesn't have a real title or description - it's not clear what is changing (or why), and what was broken before (and fixed by) this issue. It currently doesn't meet our quality guidelines to be merged so it should not be LGTM.

(Of course, the author may improve it based on our feedback, but it's not there yet)

@nancyflutter Can you:

- Edit the default PR description, indicating what you're doing and why
- Link to any issues you are fixing
- If you're not fixing an issue, explain in more detail why you're making this change
- Update the PR title as a 1-sentence summary

Thank you for your contribution.

nancyflutter committed:
Merge branch 'master' into snackbar_a11y_fix

nancyflutter committed:
Merge branch 'master' into snackbar_a11y_fix

nancyflutter committed:
Fixed all identified issues in the code:

Local variables marked as final to ensure better code practices.

Removed the undefined FlutterView class reference.

Replaced sendAnnouncement with SemanticsService.announce() for accessibility announcements.

Added type annotations where necessary, particularly for the log variable in the test file.

Correctly cast dynamic objects to Map<String, dynamic> in the test to resolve type issues

nancyflutter committed:
Replaced sendAnnouncement with SemanticsService.announce() for accessibility announcements.

nancyflutter committed:
fix: resolve analyzer warnings in snack_bar_test.dart

- Added explicit type annotations to comply with `always_specify_types`
- Avoided dynamic method call by explicitly casting and typing map entries
- Fixes `avoid_dynamic_calls` analyzer lint

shubh2294 committed:
chore: test commit to validate contribution visibility

shubh2294 committed:
chore: test commit to validate contribution visibility

nancyflutter committed:
fix: resolve analyzer warnings

nancyflutter committed:
Merge branch 'master' into snackbar_a11y_fix

nancyflutter committed:
Merge branch 'master' into snackbar_a11y_fix

nancyflutter committed:
Merge branch 'master' into snackbar_a11y_fix

nancyflutter committed:
Merge branch 'master' into snackbar_a11y_fix

loic-sharma committed:
Update dev/a11y_assessments/test/snack_bar_test.dart

Co-authored-by: chunhtai <47866232+chunhtai@users.noreply.github.com>

chunhtai commented:
the test failure seems related

```
00:15 +54: /b/s/w/ir/x/w/flutter/dev/a11y_assessments/test/snack_bar_test.dart: SnackBar Accessibility Tests snack bar announces message
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following _TypeError was thrown running a test:
type '_Map<Object?, Object?>' is not a subtype of type 'Map<String, dynamic>' in type cast

When the exception was thrown, this was the stack:
#0      main.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/dev/a11y_assessments/test/snack_bar_test.dart:37:61)
#1      ListBase.firstWhere (dart:collection/list.dart:126:15)
#2      main.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/dev/a11y_assessments/test/snack_bar_test.dart:36:13)
<asynchronous suspension>
#3      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:192:15)
<asynchronous suspension>
#4      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1059:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

The test description was:
  snack bar announces message
════════════════════════════════════════════════════════════════════════════════════════════════════
00:15 +54 -1: /b/s/w/ir/x/w/flutter/dev/a11y_assessments/test/accessibility_guideline_test.dart: testing accessibility guideline for Card
00:15 +54 -1: /b/s/w/ir/x/w/flutter/dev/a11y_assessments/test/snack_bar_test.dart: SnackBar Accessibility Tests snack bar announces message [E]
  Test failed. See exception logs above.
```

chunhtai commented:
Hi @nancyflutter are you still working on this pr?
